subsys/pm/device_runtime.c

```diff
@@ -45,11 +45,12 @@ static int runtime_suspend(const struct device *dev, bool async)
 {
 	int ret = 0;
 	struct pm_device *pm = dev->pm;
+	k_spinlock_key_t key;
 
 	if (k_is_pre_kernel()) {
 		async = false;
 	} else {
-		(void)k_mutex_lock(&pm->lock, K_FOREVER);
+		key = k_spin_lock(&pm->lock);
```

Comment:
since the `pm_action_callback` can potentially preempt (e.g. due to an I2C/SPI transfer), wouldn't it be better to use semaphores, which can be used in ISR context?

Reply:
With the semaphore in the ISR I cannot wait if the semaphore is already taken. In that case we should return something like `-EAGAIN` to the user and that would (probably?) break the current users of this API.

Reply:
Note that we should only modify the `put_async` behavior, by returning -EAGAIN if lock can't be obtained in ISR context. I guess it's the best we can do. Note that PM API is unstable, and I'd consider this a fix really.